### ISSUE TITLE
Excessive codegen time for zip with large tuple elements

### ISSUE THREAD

**Issue by martinholters** [JuliaLang/julia#35847]
```julia
julia> z = zip([ntuple(identity, 320)]);

julia> @time code_llvm(devnull, collect, typeof((z,)))
  0.131978 seconds (289.26 k allocations: 15.942 MiB)

julia> @time code_native(devnull, collect, typeof((z,)))
 46.796472 seconds (6.56 k allocations: 329.421 KiB)

julia> # note the bad the scaling, too

julia> @time code_native(devnull, collect, typeof((zip([ntuple(identity, 350)]),)))
 72.477331 seconds (174.13 k allocations: 9.846 MiB, 0.01% gc time)
```
Note that it does need the `Zip` layer to trigger the long codegen.
Similar behavior on 1.6.0-DEV.17 and 1.5.0-DEV.876. On 1.4.1, it is much worse, so at least there has been some progress.

**Comment by georgematheos** [comment]
I'm encountering a similar issue; I don't fully understand why it's happening, but wanted to add a quick profile dump in case this is helpful for others.

My use case is when using zip over a large number of short iterators.  For example,
```julia
itrs = [(i, i + 1) for i=1:10]
z = zip(itrs...)
collect(z)
```
is quite slow (and when the length of `itrs` grows, this gets much much slower--I think worse than linearly.)  It looks like the slowness all comes from the compilation; when I run the collect line again, it runs in only a few milliseconds.

Here is dump of a profile of the compilation (calling `@profile collect(z)` in the example above, when running `collect` for the first time): https://gist.github.com/georgematheos/e2b6a238b1c16cd4401973cb6a719792

**Comment by StefanKarpinski** [comment]
Don't splat things that can grow arbitrarily. What you need here is [the `unzip` function](https://github.com/JuliaLang/julia/pull/33515), which unfortunately hasn't been merged yet due to bickering over performance.

**Comment by o314** [comment]
Isn't it closed since v1.7 ?

```
julia> VERSION
v"1.7.1"

julia> z = zip([ntuple(identity, 320)]);

julia> @time code_llvm(devnull, collect, typeof((z,)))
  0.139754 seconds (483.75 k allocations: 26.299 MiB, 61.14% compilation time)

julia> @time code_native(devnull, collect, typeof((z,)))
  0.689610 seconds (5.97 k allocations: 330.344 KiB, 87.96% compilation time)

julia> @time code_native(devnull, collect, typeof((zip([ntuple(identity, 350)]),)))
  1.492174 seconds (427.12 k allocations: 23.065 MiB, 0.71% gc time, 99.89% compilation time)
```

But not on v1.6

```
julia> VERSION
v"1.6.3"

julia> z = zip([ntuple(identity, 320)]);

julia> @time code_llvm(devnull, collect, typeof((z,)))
  0.267153 seconds (582.18 k allocations: 33.995 MiB, 33.57% gc time, 77.29% compilation time)

julia> @time code_native(devnull, collect, typeof((z,)))
 32.379750 seconds (5.93 k allocations: 356.084 KiB, 99.94% compilation time)

julia> @time code_native(devnull, collect, typeof((zip([ntuple(identity, 350)]),)))
 47.863262 seconds (449.59 k allocations: 26.478 MiB, 100.00% compilation time)
```

**Comment by vtjnash** [comment]
cool. we should potentially add it as a BaseBenchmarks.jl to nanosoldier, but we can close
